### PR TITLE
Matomo integration

### PR DESCRIPTION
**Pull request by FabianDevel**
Integrate matomo app tracking in kdrive

- track view changes
- user interaction with the app (click on buttons)

### CODE REVIEW

File: app/src/main/java/com/infomaniak/drive/ApplicationMain.kt [hunk 1]
```diff
@@ -54,13 +61,18 @@ import io.sentry.SentryEvent
...
 import okhttp3.Interceptor
 import okhttp3.OkHttpClient
-import org.eclipse.paho.client.mqttv3.*
+import org.matomo.sdk.Tracker
+import org.matomo.sdk.extra.MatomoApplication
 import java.util.*
 
-class ApplicationMain : Application(), ImageLoaderFactory {
+
+class ApplicationMain : MatomoApplication(), ImageLoaderFactory {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ApplicationMain.kt:75]: Is it possible to do it another way?

It's generally a bad idea to use another Application class. The day we'll need to do the same for another library, we'll be blocked.

File: app/src/main/java/com/infomaniak/drive/utils/TabViewPagerUtils.kt [hunk 2]
```diff
@@ -48,7 +49,7 @@ object TabViewPagerUtils {
             if (isChecked) {
                 val position = (tabsViewPager.adapter as ViewPagerAdapter).tabs.indexOfFirst { it.button == checkedId }
                 tabsViewPager.setCurrentItem(position, true)
-                onCheckedButton?.invoke(position)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/utils/TabViewPagerUtils.kt:51]: Why did you remove this line?

Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/utils/TabViewPagerUtils.kt:51]: Probably a bad selection on paste, good catch

File: app/src/main/java/com/infomaniak/drive/ApplicationMain.kt (hunk 1 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/ApplicationMain.kt]: I've change it with a lazy for the tracker instanciation, thanks for the tip

File: app/src/main/java/com/infomaniak/drive/utils/MatomoUtils.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import org.matomo.sdk.Matomo
+import org.matomo.sdk.Tracker
+import org.matomo.sdk.TrackerBuilder
+import org.matomo.sdk.extra.DownloadTracker
+import org.matomo.sdk.extra.TrackHelper
+
+object MatomoUtils {
+
+    fun Context.buildTracker(): Tracker {
+        return TrackerBuilder("https://analytics.infomaniak.com/matomo.php", 8, "AndroidTracker").build(Matomo.getInstance(this))
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/utils/MatomoUtils.kt:36]: Add this api url in Gradle

File: app/src/main/java/com/infomaniak/drive/ui/BaseActivity.kt [hunk 4]
```diff
@@ -30,5 +31,6 @@ open class BaseActivity : AppCompatActivity() {
         if (AccountUtils.currentUser == null) {
             runBlocking { AccountUtils.requestCurrentUser() }
         }
+        trackScreen(this::class.java.name.substringAfter("com.infomaniak.drive."))
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/BaseActivity.kt:34]:
```suggestion
        trackScreen(this::class.java.name.substringAfter("${BuildConfig.APPLICATION_ID}."))
```

File: app/src/main/java/com/infomaniak/drive/utils/MatomoUtils.kt [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import org.matomo.sdk.Tracker
+import org.matomo.sdk.TrackerBuilder
+import org.matomo.sdk.extra.DownloadTracker
+import org.matomo.sdk.extra.TrackHelper
+import java.util.*
+
+object MatomoUtils {
+
+    private inline val Context.tracker: Tracker get() = (this as ApplicationMain).matomoTracker
+    private const val siteId = 8
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/utils/MatomoUtils.kt:37]: Magic number.

File: app/src/main/java/com/infomaniak/drive/ui/LaunchActivity.kt [hunk 6]
```diff
@@ -40,29 +42,28 @@ class LaunchActivity : AppCompatActivity() {
...
                 }
                 isKeyguardSecure() && AppSettings.appSecurityLock -> {
-                    startActivity(Intent(this@LaunchActivity, LockActivity::class.java))
+                    LockActivity::class.java
                 }
                 else -> {
                     if (DriveInfosController.getDrivesCount(AccountUtils.currentUserId) == 0L) {
                         AccountUtils.updateCurrentUserAndDrives(this@LaunchActivity)
                     }
+                    application.trackCurrentUserId()
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/LaunchActivity.kt:56]: Move to first line in else case

File: app/src/main/java/com/infomaniak/drive/ui/LaunchActivity.kt [hunk 7]
```diff
@@ -40,29 +42,28 @@ class LaunchActivity : AppCompatActivity() {
 
             logoutCurrentUserIfNeeded() // Rights v2 migration temporary fix
 
-            when {
+            val clazz = when {
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/LaunchActivity.kt:45]: Rename to class

Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/ui/LaunchActivity.kt:45]: I can't do it as class is a reserved keyword

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt [hunk 8]
```diff
@@ -60,19 +60,20 @@ class PreviewSliderFragment : Fragment(), FileInfoActionsView.OnItemClickListene
 
     private val mainViewModel: MainViewModel by activityViewModels()
     private val previewSliderViewModel: PreviewSliderViewModel by navGraphViewModels(R.id.previewSliderFragment)
+    private val selectFolderResultLauncher = registerForActivityResult(StartActivityForResult()) {
+        it.whenResultIsOk { data -> onSelectFolderResult(data) }
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt:65]: Revert this change.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt [hunk 9]
```diff
@@ -60,19 +60,20 @@ class PreviewSliderFragment : Fragment(), FileInfoActionsView.OnItemClickListene
...
 
     private lateinit var bottomSheetBehavior: BottomSheetBehavior<View>
-    private lateinit var currentPreviewFile: File
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewSliderFragment.kt:65]: Revert this change.

**Review by JorisBodin — APPROVED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Review by sirambd — APPROVED** [review]
